[28기 박준영] 리뷰 반영 및 오류 수정

joonyg10 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [X] 기능 추가
- [X] 리뷰 반영
- [X] 리팩토링
- [ ] 버그 수정
- [X] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 인스타그램과 최대한 비슷하도록 로그인 및 Main 페이지 구현 

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
**1. 로그인 페이지 기능 구현**
  - 로그인 페이지 레이아웃 작성
  - email과 password에 입력하는 경우 각각 handleEmail, handlePwd 함수가 시행되어 입력한 값이 email, password라는 state에 저장
  - email과 password가 수정되는 경우 useEffect를 이용하여 입력에 대한 유효성 검사 (validation) 시행
  - 위의 유효성 검사값을 active라는 state로 저장하였고 active가 true인 경우 로그인 버튼에 active라는 class를 주어 색상 변경
  - input창에서 enter 또는 로그인 버튼 클릭시 유효성 검사를 통과한 경우에 한해 메인 페이지로 이동

**2. 메인페이지** 
  - 메인페이지의 경우 크게 Story, Sidebar, Feed 3개의 컴포넌트로 분리하였고 각 컴포넌트에 대한 구현 설명은 아래 작성. (3 ~ 5)
  - 레이아웃은 grid속성을 사용!
  - 부수적으로 Heart (하트 버튼), Sidebar 등에서 반복되는 부분은 따로 컴포넌트로 분리
  - fontAwesome, flaticon대신 react-icons 라이브러리를 사용하여 icon을 사용
  - 네비게이션 바 제외, 모든 이미지에 로딩 효과 스타일 추가

**3. Story 컴포넌트** 
 - 각 스토리에 대해 그라데이션이 추가된 선 스타일링
 - 양쪽의 버튼을 클릭하는 경우 스토리 부분이 슬라이드 되도록 구현
 - 각 스토리 이미지를 클릭하는 경우 해당 이미지와 이름을 가지는 Modal 띄우는 기능을 구현 중  

**3-1) Modal**
  - Modal 컴포넌트 생성
  - story의 이미지를 클릭 했을 때, 해당 story의 이미지와 username으로 Modal을 업데이트
  - Modal의 welcome 메시지에 하트가 올라오는 애니메이션 기능 추가 진행 중

**4. Sidebar 컴포넌트** 
  - position fixed로 스크롤 되어도 같은 곳에 위치하도록 함
  - '회원님을 위한 추천' 밑의 부분 또한 똑같은 구조였기에 Others 라는 컴포넌트를 따로 생성해 중복 최소화
  - other에 해당하는 부분을 data.json으로 이동
  
**5. Feed 컴포넌트** 
   - Feed의 컴포넌트도 header, Carousel (이미지 슬라이더 ), 아이콘, 댓글로 4부분으로 컴포넌트 분리를 시행하였다.
   - grid와 grid-template-rows 속성을 이용하여 각 영역을 구분
   
**5-1 ) Feed 이미지 슬라이더**
   - 5개의 이미지를 사용하여 캐러셀 생성하였고 현재 보여줄 이미지를 관리하기 위해 image의 index를 state로 관리
   - 슬라이더의 기본 틀이 되는 요소에 useRef를 이용하여 refenrence를 연동하였음.  ( = carouselRef ...1) )
   - 이미지 양쪽의 버튼을 클릭시 왼쪽 또는 오른쪽으로 이동하며 처음,  마지막 인덱스에 도달하는 경우 button에 display: none; 스타일 적용
   - 버튼 클릭시 인덱스를 업데이트 하도록 setState를 호출하였으며 useEffect에서 deps로 index를 주고 side effect로 1)의 style.transform = `translateX(- ${idx * 560}px )`으로 이미지 이동하는 방식을 구현 ( 560px은 feed의 width )
   - 캐러셀 밑에 작은 네비게이터도 이미지의 갯수에 맞게 생성하였으며 현재 이미지에 해당하는 index에 .current라는 클래스를 주어 파란 색상을 부여
   - 이미지를 슬라이드 할 때, 네비게이터 또한 해당 index로 자동으로 업데이트 되며 네비게이터 클릭시 해당하는 index의 이미지로 바로 이동하도록 구현 

**5-2) FeedIcons 컴포넌트** 
  - 피드의 캐러셀 밑의 아이콘들과 좋아요 수 표시를 담당하는 컴포넌트 
  - 좋아요 수는 부모 컴포넌트인 Feed에서 관리하며 props로 받아 .toLocaleString() 메소드로 표시
  - 6)에서 설명할 Heart 컴포넌트를 불러와 구현하였음.

**5-3) FeedComment 컴포넌트**
   - 댓글 표시, 추가, 삭제를 담당하는 UI 컴포넌트
   - 댓글 전체는 comments라는 state내에 댓글에 대한 정보를 배열 형태로 관리하였고 댓글 하나하나는 Comment라는 컴포넌트로 분리해 ul 내부에 댓글.map으로 
   - 댓글에 대한 정보는 { id , username, comment } 가 state로 관리되며 Comment의 props로 전달
   - 댓글을 왼쪽으로 드래그 했을 때, 밑에 시행취소, 댓글 삭제 버튼이 나타나도록 requestAnimationFrame을 사용해 설계했으며 너무 많이 드래그 하여 벗어나면 안되기에 이동거리에 제한을 걸어주었다.
   - 시행취소 (X) 버튼 클릭 시 translate 초기화
   - 삭제 버튼 클릭 시 댓글 삭제 되도록 구현 

**6) Heart 컴포넌트**
  - 5-2)의 FeedIcons와 5-3)의 FeedComment 컴포넌트에 같이 사용될 부분이고 로직이 길어져 따로 컴포넌트로 분리
  - 자신이 click되었는지에 대한 boolean을 state로 저장하며 true / false에 따라 다른 heart를 rendering
  - click시 Feed의 좋아요 수가 업데이트 되는 함수를 props로 받고 시행
  - 클릭시 심장이 뛰는 듯한 애니메이션 추가 
  
**7) Mock data**
  - public/data/joonyoung이란 디렉토리 생성 및 data.json이라는 파일 생성
  - Main.jsx에서 useEffect, fetch를 활용해 data.json의 내용을 가져와 Main.jsx에 state로 저장 
  - 위의 저장한 데이터를 각 컴포넌트의 props로 전달하여 수정


<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- React에서 컴포넌트를 좀 더 효율적으로 나누는 방식을 생각하게 되었습니다.
- css custom property를 사용해 기존의 css에서 반복해서 작성하는 부분을 최소화 할 수 있었습니다. (로딩 효과)
- git으로 프로젝트가 어떤 형태로 흘러가는지에 대한 윤곽이 조금 보이게 되었습니다.

<br />

## :: 기타 질문 및 특이 사항
- 제가 setState를 사용할 때 습관적으로 setState(prev => prev + 1), setState(prev => !prev) 과 같은 형태로 자주 사용합니다. 혹시 이렇게 쓰는게 문제가 있는지, 아니면 굳이 이전 상태값이 필요없는 경우 이렇게 쓰는 것보단 바로 setState(newValue)형태로 쓰는게 더 좋은지 궁금합니다. 
- 이미지 크기를 조정할 때, width, height를 둘다 고정값을 주는 방식보다는 min-width, max-width를 지정하고 height만 고정하는 방식으로 프로젝트를 진행했습니다. 둘다 고정된 값을 주면 실제 이미지 비율이 깨져서 이렇게 했습니다만 혹시 더 좋은 방법이 있을까요?

joonyg10 committed:
[ADD] 로그인 페이지 구현
사용할 이미지들 추가
비밀번호 잊으셨나요 클릭시 instagram 비밀번호 찾기 페이지로 이동

joonyg10 committed:
[ADD] 사이드바 부분 메인페이지에 추가
[FIX] 스토리 부분 center안맞춰진 부분 및 이미지 크기 수정
[ADD] 스토리 마다 그라데이션 색상의 원을 p추가

joonyg10 committed:
Merge branch 'master' of https://github.com/wecode-bootcamp-korea/28-React-Westagram-2 into feature/joonyoung

joonyg10 committed:
[ADD] 피드 캐러셀 부분 추가 중

joonyg10 committed:
Merge branch 'master' into feature/joonyoung
네비게이션 및 mixin 수정 사항 반영

joonyg10 committed:
[FIX] 피드 캐러셀

joonyg10 committed:
[ADD] 피드의 캐러셀 부분 리액트로 바꾸어 추가

joonyg10 committed:
[ADD] 피드의 아이콘 영역 추가

joonyg10 committed:
[ADD] 피드의 댓글 입력 부분 레이아웃, 댓글 추가기능 구현
[ADD] 댓글과 icon부분을 컴포넌트를 쪼개었기에 두 컴포넌트에서 모두 사용하도록 Heart라는 컴포넌트 생성

joonyg10 committed:
[ADD] 댓글을 왼쪽으로 드래그 했을 때, 움직이며 아래의 취소, 삭제 버튼이 나오도록 구현
[ADD] 댓글 추가 버튼이 댓글이 있을 때만 (active) 하도록 구현

joonyg10 committed:
[ADD] mock.data활용을 위해 public/data/joonyoung 디렉토리 내에 feed.json 생성
[ADD] 위의 feed.json을 fetch해온 데이터를  Main 컴포넌트에 추가
[FIX] Main으로 가져온 데이터를 하위 컴포넌트들의 props로 전달해 주어 컴포넌트 별로 표기되는 내용들 수정
[ADD] 피드의 이미지 슬라이더에 로딩 효과 추가
[Refactor] 중복된 img 태그의 내용 main.scss에 img로 단일화 및 이미지, feedHeader의 padding 조절

[RefactoR] Comment 컴포넌트가 댓글 입력창에 무언가를 작성할 때마다 re-rendering되어 react.memo로 불필요한 re-render 최소화

[ADD] 댓글 입력 창에서 shift + enter 누르는 경우 댓글이 추가되지 않고 새로운 줄이 생성되도록 하는 기능 추가

joonyg10 committed:
[ADD] story의 이미지 클릭 시 Modal 컴포넌트가 나타나도록 구현 및 Modal 컴포넌트 생성

YOONKY commented on public/data/joonyoung/data.json:
```diff
@@ -0,0 +1,71 @@
+      ]
+    }
+  ]
+}
```
mockData king 준영님👍 정말 감사합니다. 세부 내용은 임의로 변경한 후, 데이터의 형식좀 빌려다 쓰겠습니다 감사합니다!

YOONKY commented on src/pages/joonyoung/Login/Login.jsx:
```diff
@@ -1,5 +1,87 @@
+    navigate('/main-joonyoung');
+  };
+
+  useEffect(() => {
+    validateInput();
+  }, [email, password]);
```
이 부분을 왜 다시 유즈이펙트로 넣어주신 건지 궁금합니다!

joonyg10 replied:
email과 password가 바뀔 때마다 validation을 시행하도록 생각을 했습니다. 그런데 이렇게 할 필요가 없었더라고요....

YOONKY reviewed (commented):
감히 제가 무엇을 논하겠습니까? 감사합니다.

joonyg10 committed:
[ADD] Modal창 추가 및 하트 올라오는 애니메이션 추가
[Refactoring] scss 파일 리펙토링

bonbon0202 commented on src/pages/joonyoung/Login/Login.jsx:
```diff
@@ -1,5 +1,88 @@
+
+  const goToMainByEnter = e => {
+    if (e.key === 'Enter') goToMain();
+  };
```
함수명 잘 지어주셨습니다! good!

bonbon0202 commented on src/pages/joonyoung/Main/Feed/Feed.jsx:
```diff
@@ -0,0 +1,38 @@
+import FeedCarousel from './FeedCarousel/FeedCarousel';
+import './Feed.scss';
+import FeedIcons from './FeedIcons/FeedIcons';
+import FeedComment from './FeedComment/FeedComment';
```
import 순서 컨벤션에 맞게 수정해주세요! 사소한 부분이라 놓치기 쉽지만 이런 컨벤션 준수들이 모여서 좀 더 가독성 좋은 코드가 됩니다. week4에 안내해드린 [Refactoring Checklist](https://www.notion.so/Westagram-React-Refactoring-Checklist-9422c31c960547f3939fcf7f6c7c0014) 및 아래 순서 참고해주세요.

- 라이브러리
    - React 관련 패키지
    - 외부 라이브러리
- 컴포넌트
    - 공통 컴포넌트 → 먼 컴포넌트 → 가까운 컴포넌트
- 함수, 변수 및 설정 파일
- 사진 등 미디어 파일(`.png`)
- css 파일 (.`scss`)

bonbon0202 commented on src/pages/joonyoung/Main/Feed/Feed.scss:
```diff
@@ -0,0 +1,12 @@
+.feed {
+  margin-top: 1.75rem;
+  display: grid;
+  grid-template-rows: min-content 60vh max-content auto;
+  box-shadow: 0 0 0.75rem lightgray;
+  position: relative;
+  font-family: sans-serif;
```
css 속성 순서
- Layout Properties (position, float, clear, display)
- Box Model Properties (width, height, margin, padding)
- Visual Properties (color, background, border, box-shadow)
- Typography Properties (font-size, font-family, text-align, text-transform)
- Misc Properties (cursor, overflow, z-index)

joonyg10 replied:
리뷰 감사합니다!! 혹시 transform, transition, animation 속성을 쓰게 된다면 Misc properties 밑에 쓰면 컨벤션에 맞을까요?

bonbon0202 commented on src/pages/joonyoung/Main/Feed/FeedCarousel/FeedCarousel.jsx:
```diff
@@ -0,0 +1,71 @@
+
+const FeedCarousel = ({ carouselImages }) => {
+  const [currIdx, setCurrIdx] = useState(0);
+  const carouselRef = useRef(null);
```
ref 잘이용해주셨네요! 좋습니다

bonbon0202 commented on src/pages/joonyoung/Main/Feed/FeedCarousel/FeedCarousel.jsx:
```diff
@@ -0,0 +1,71 @@
+          size="24"
+          className={`arrow ${
+            currIdx === carouselImages.length - 1 ? 'hidden' : ''
+          }`}
```
템플릿 리터럴 잘 활용해주셨네요! 위에 리뷰 달아드렸듯이, 조건에 대해 가독성을 높이기 위해 변수를 활용하는 것도 좋을 것 같습니다! :)

bonbon0202 commented on src/pages/joonyoung/Main/Feed/FeedIcons/FeedIcons.jsx:
```diff
@@ -0,0 +1,22 @@
+const FeedIcons = ({ likes, setLikes }) => {
+  return (
+    <section className="feed__icons">
+      <article>
+        <Heart size="20" setLikes={setLikes} />
+        <FaRegComment size="20" />
+        <FaRegPaperPlane size="20" />
+        <FaRegBookmark size="20" />
+      </article>
```
article은 따로 떼어내어 사용해도 무방할 만큼 독립적인 기사문등과 같은 요소에 주로 사용합니다. 해당 요소를 article로 하신 이유가 있으실까요?!

joonyg10 replied:
article 부분이 아이콘들만 묶여있고 그 밑에 오는 <p> 좋아요 : {likes} <p/>와 구분하려고 사용하였습니다. 시멘틱 배운 후 최대한 div안 쓰고 section내에서 구역을 나누려고 해서 사용한 것 같습니다. 수정하고 시멘틱 태그 조금더 살펴보겠습니다. 감사합니다!

bonbon0202 commented on src/pages/joonyoung/Main/Feed/Heart/Heart.jsx:
```diff
@@ -0,0 +1,28 @@
+import { AiFillHeart, AiOutlineHeart } from 'react-icons/ai';
+import './Heart.scss';
+
+const Heart = ({ size, setLikes }) => {
+  const [clicked, setClicked] = useState(false);
+
+  const handleClickHeart = () => {
+    setClicked(prev => !prev);
+    setLikes(prevLikes => (clicked ? prevLikes - 1 : prevLikes + 1));
+  };
+
+  return clicked ? (
+    <AiFillHeart
+      className={`heart ${clicked ? 'like' : ''}`}
+      size={size}
+      onClick={handleClickHeart}
+    />
+  ) : (
+    <AiOutlineHeart
+      className={`heart ${clicked ? 'like' : ''}`}
+      size={size}
+      onClick={handleClickHeart}
+    />
+  );
+};
```
clicked 값에 따라서 이미 다른 모양의 아이콘이 렌더링되는 것 같은데 className에서 한번 더 조건을 작성해주신 이유가 있으실까요?

joonyg10 replied:
처음에 작성했던걸 그대로 가져오느라 빠뜨린 것 같습니다. 바로 수정하겠습니다.

bonbon0202 commented on src/pages/joonyoung/Main/Main.jsx:
```diff
@@ -1,5 +1,53 @@
+    fetchFeed();
+  }, []);
+
+  if (feedDatas.length === 0) return <h1>Loading datas...</h1>;
```
오우 로딩까지 핸들링해주셨네요! ㅎㅎ 좋습니다

bonbon0202 commented on src/pages/joonyoung/Main/Modal/Modal.jsx:
```diff
@@ -0,0 +1,32 @@
+    return () => {
+      window.removeEventListener('keydown', e => closeModal(e.key));
+    };
+  }, []);
```
클린업 사용까지! 잘 사용해주셨네요!

bonbon0202 commented on src/pages/joonyoung/Main/Sidebar/Sidebar.jsx:
```diff
@@ -0,0 +1,42 @@
+    { id: 1, username: 'wecode28기', description: 'wecode님외 36명이 follow' },
+    { id: 2, username: 'Younha', description: 'Enna Kim외 10명이 follow' },
+    { id: 3, username: 'wework', description: 'wecode28님외 236명이 follow' },
+  ];
```
위와 같이 함수에 영향을 받지 않는 함수나 변수들은 불필요한 메모리할당을 줄이기 위해 함수형 컴포넌트 밖에 선언해주시는 것이 좋습니다. 더불어 위와 같이 변하지 않는 상수 데이터는 아래와 같이 '대문자'와 '_'을 이용해 네이밍하는 것이 컨벤션입니다.
```suggestion
  const OTHER_PROFILE_INFOS = [
    { id: 1, username: 'wecode28기', description: 'wecode님외 36명이 follow' },
    { id: 2, username: 'Younha', description: 'Enna Kim외 10명이 follow' },
    { id: 3, username: 'wework', description: 'wecode28님외 236명이 follow' },
  ];
```

joonyg10 committed:
[Refactoring] scss feed, story, modal refactoring

bonbon0202 commented on src/pages/joonyoung/Main/Feed/FeedCarousel/FeedCarousel.jsx:
```diff
@@ -0,0 +1,70 @@
+    setCurrIdx(prevIdx => --prevIdx);
+  };
+
+  const handleSlideRight = () => {
+    if (currIdx === 4) return;
+    setCurrIdx(prevIdx => ++prevIdx);
+  };
```
index가 마지막일 때 리턴을 해주시는 것 같네요! 숫자 4의 의미가 드러날 수 있도록 변수에 할당에 활용하는 것도 방법입니다.

```suggestion
  const handleSlideRight = () => {
    const lastCardIndexNumber = 4;
    if (currIdx === lastCardIndexNumber) return;
    setCurrIdx(prevIdx => ++prevIdx);
  };
```

joonyg10 replied:
!!!! 팁 감사합니다!!

bonbon0202 commented on src/pages/joonyoung/Main/Feed/FeedComment/Comment/Comment.jsx:
```diff
@@ -0,0 +1,85 @@
+    frontRef.current.style.transform = `translateX(-${movedBy}px)`;
+  };
+
+  // useEffect(() => {
+  //   if (movedX === 0) return;
+  //   let movedBy = startX - movedX;
+  //   if (movedBy >= 50) movedBy = 100;
+  //   frontRef.current.style.transform = `translateX(-${movedBy}px)`;
+  // }, [movedX]);
```
push 전에 삭제해야할 주석은 없는지 항상 확인해주세요!

joonyg10 committed:
[Refactoring] Login페이지 불필요한 연산, state삭제
[Refactoring] sidebar에서 사용되는 데이터 data.json으로 이동동
[Fix] Main에서 story, feed에 대한 데이터를 2개의 state로 관리한 것 1개로 통일

joonyg10 committed:
[Refactoring] Login, main페이지의 컨벤션 맞추기
[FIX] feedCarousel, comment에서 useRef를 사용한 부분을

joonyg10 committed:
[Refactoring] 로그인, 메인 페이지 컨벤션 적용
[FIX] story, feedCarousel, comment에서 useRef로 DOM에 접근하는 방식 대신 css custom property를 이용한 방식으로 수정

joonyg10 committed:
[Refactoring] Modal, main에서 모달에 관한 css스타일링 리펙토링
